src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "flight/position.h"
+#include "flight/imu.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 1);
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+);
+
+
+typedef struct {
+    float max;
+    float min;
+    float kp;
+    float kd;
+    float ki;
+    float lastErr;
+    float integral;
+} simplePid_s;
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+typedef struct {
+    simplePid_s altPid;
+    simplePid_s velPid;
+    float throttle;
+    float targetAltitude;
+    float measuredAltitude;
+    float measuredAccel;
+    float velocityEstimationAccel;  // based on acceleration
+    float startVelocityEstimationAccel;
+    bool prevAltHoldModeEnabled;
+} altHoldState_s;
+
+void altHoldReset(altHoldState_s* altHoldState)
```

Comment:
Maybe: Pass target altitude as parameter

Reply:
Don't see why it can be better. Reset called from two different places.

Reply:
You are already mixing initialization, parameter update (but in-flight tuning is not implemented) and PID state reset.
But setpoint update will be necessary soon, so this refactoring can be left for later.